README.md
# Cursor Locator #
* Author: Noelia Ruiz Martínez.
* Compatibility: 2020.4.

This add-on makes possible to know the position of the system caret respect to the start of the current line, while typing to add text in multiline controls.

## Cursor Locator settings ##

This panel is available from NVDA's menu, Preferences submenu, Settings dialog.

It provides the following options:

* Report start of line: When this control is checked, a low tone will announce if the caret is at the start of the current line while text is been added. (Checked by default).
* Report line length: You can type or choose a line length (number of characters between 0 and 600), which will be announced by a hight tone when it's reached. (The default value is 80 characters).
* Pitch of sound for start of line: You can type or select a value between 20 and 20000. (The default value is 400 hertzs).
* Length of sound for start of line: You can type or select a value between 20 and 2000. (The default value is 50 milliseconds).
* Test sound for start of line: Press this button to test the configured sound for start of line.
* Pitch of sound for end of line: You can type or select a value between 20 and 20000. (The default value is 1000 hertzs).
* Length of sound for end of line: You can type or select a value between 20 and 2000. (The default value is 50 milliseconds).
* Test sound for end of line: Press this button to test the configured sound for end of line.

## Commands ##

You can modify the gestures to the following commands, trought the NVDA's menu > Preferences submenu > Input gestures dialog.

* NVDA+control+shift+l: When possible, reports the lenght of the current line (System caret category).
* Not assigned: Shows the Cursor Locator settings dialog (Config category).

## Changes for 1.0 ##
* Initial version



changelog.md
## Changes for 1.5.1-dev (PR #7) ##
* Added brazilian and european portuguese translations.


buildVars.py
# -*- coding: UTF-8 -*-

# Build customizations
# Change this file instead of sconstruct or manifest files, whenever possible.

# Full getext (please don't change)
_ = lambda x : x

# Add-on information variables
addon_info = {
	# for previously unpublished addons, please follow the community guidelines at:
	# https://bitbucket.org/nvdaaddonteam/todo/raw/master/guideLines.txt
	# add-on Name, internal for nvda
	"addon_name" : "cursorLocator",
	# Add-on summary, usually the user visible name of the addon.
	# Translators: Summary for this add-on to be shown on installation and add-on information.
	"addon_summary" : _("Cursor Locator"),
	# Add-on description
	# Translators: Long description to be shown for this add-on on add-on information from add-ons manager
	"addon_description" : _("""Reports cursor positions while typing on multiline edit controls."""),
	# version
	"addon_version" : ,
	# Author(s)
	"addon_author" : u"Noelia Ruiz Martínez <nrm1977@gmail.com>",
	# URL for the add-on documentation support
	"addon_url" : "https://github.com/nvdaes/cursorLocator",
	# Documentation file name
	"addon_docFileName" : "readme.html",
	# Minimum NVDA version supported (e.g. "2018.3")
	"addon_minimumNVDAVersion" : "2020.4",
	# Last NVDA version supported/tested (e.g. "2018.4", ideally more recent than minimum version)
	"addon_lastTestedNVDAVersion" : "2020.4",
	# Add-on update channel (default is stable or None)
	"addon_updateChannel" : None,
}


import os.path

# Define the python files that are the sources of your add-on.
# You can use glob expressions here, they will be expanded.
pythonSources = [os.path.join("addon", "cursorLocator", "*.py")]

# Files that contain strings for translation. Usually your python sources
i18nSources = pythonSources + ["buildVars.py"]

# Files that will be ignored when building the nvda-addon file
# Paths are relative to the addon directory, not to the root directory of your addon sources.
excludedFiles = []


addon/globalPlugins/cursorLocator/__init__.py
# -*- coding: UTF-8 -*-

# cursorLocator: Global plugin to know the cursor position when typing on multiline edit controls
# Copyright (C) 2017-2021 Noelia Ruiz Martínez
# Released under GPL 2

import addonHandler
import globalPluginHandler
import controlTypes
import textInfos
import treeInterceptorHandler
import api
import ui
import tones
import config
import wx
import gui
from gui import SettingsPanel, NVDASettingsDialog, guiHelper, nvdaControls
from scriptHandler import script
from globalCommands import SCRCAT_SYSTEMCARET, SCRCAT_CONFIG

addonHandler.initTranslation()

# Constants
ADDON_SUMMARY = addonHandler.getCodeAddon().manifest["summary"]
ADDON_PANEL_TITLE = ADDON_SUMMARY

confspec = {
	"reportStartOfLine": "boolean(default=True)",
	"reportLineLength": "integer(default=80)",
	"startLinePitch": "integer(default=400)",
	"startLineLength": "integer(default=50)",
	"endLinePitch": "integer(default=1000)",
	"endLineLength": "integer(default=50)",
}
config.conf.spec["cursorLocator"] = confspec


class AddonSettingsPanel(SettingsPanel):

	title = ADDON_PANEL_TITLE

	# Translators: Description of the Cursor Locator panel.
	panelDescription = _("Configure {}").format(ADDON_SUMMARY)

	def makeSettings(self, settingsSizer):
		sHelper = guiHelper.BoxSizerHelper(self, sizer=settingsSizer)
		sHelper.addItem(wx.StaticText(self, label=self.panelDescription))

		# Translators: Label for a group of Cursor Locator options.
		lineGroupText = _("Line properties")
		lineGroup = guiHelper.BoxSizerHelper(
			self, sizer=wx.StaticBoxSizer(wx.StaticBox(self, label=lineGroupText), wx.VERTICAL)
		)
		sHelper.addItem(lineGroup)

		# Translators: Label for the Cursor Locator panel.
		self.reportStartCheckBox = lineGroup.addItem(wx.CheckBox(self, label=_("&Report start of line")))
		self.reportStartCheckBox.SetValue(config.conf["cursorLocator"]["reportStartOfLine"])

		# Translators: Label for the Cursor Locator panel.
		self.LengthEdit = lineGroup.addLabeledControl(
			_("Report &line length:"), nvdaControls.SelectOnFocusSpinCtrl,
			min=0, max=600, initial=config.conf["cursorLocator"]["reportLineLength"]
		)

		# Translators: Label for a group of Cursor Locator options.
		startGroupText = _("Sound for start of line")
		startGroup = guiHelper.BoxSizerHelper(
			self, sizer=wx.StaticBoxSizer(wx.StaticBox(self, label=startGroupText), wx.VERTICAL)
		)
		sHelper.addItem(startGroup)

		# Translators: Label for the Cursor Locator panel.
		self.startHzEdit = startGroup.addLabeledControl(
			_("&Pitch of sound for start of line:"), nvdaControls.SelectOnFocusSpinCtrl,
			min=20, max=20000, initial=config.conf["cursorLocator"]["startLinePitch"]
		)

		# Translators: Label for the Cursor Locator panel.
		self.startLengthEdit = startGroup.addLabeledControl(
			_("L&ength of sound for start of line:"), nvdaControls.SelectOnFocusSpinCtrl,
			min=20, max=2000, initial=config.conf["cursorLocator"]["startLineLength"]
		)

		# Translators: Label for the Cursor Locator panel.
		label = _("&Test sound for start of line")
		self.testStartSoundButton = startGroup.addItem(wx.Button(self, label=label))
		self.testStartSoundButton.Bind(wx.EVT_BUTTON, self.onTestStartSound)

		# Translators: Label for a group of Cursor Locator options.
		endGroupText = _("Sound for end of line")
		endGroup = guiHelper.BoxSizerHelper(
			self, sizer=wx.StaticBoxSizer(wx.StaticBox(self, label=endGroupText), wx.VERTICAL)
		)
		sHelper.addItem(endGroup)

		# Translators: Label for the Cursor Locator panel.
		self.endHzEdit = endGroup.addLabeledControl(
			_("P&itch of sound for end of line:"), nvdaControls.SelectOnFocusSpinCtrl,
			min=20, max=20000, initial=config.conf["cursorLocator"]["endLinePitch"]
		)

		# Translators: Label for the Cursor Locator panel.
		self.endLengthEdit = endGroup.addLabeledControl(
			_("Le&ngth of sound for end of line:"), nvdaControls.SelectOnFocusSpinCtrl,
			min=20, max=2000, initial=config.conf["cursorLocator"]["endLineLength"]
		)

		# Translators: Label for the Cursor Locator panel.
		label = _("Test s&ound for end of line")
		self.testEndSoundButton = endGroup.addItem(wx.Button(self, label=label))
		self.testEndSoundButton.Bind(wx.EVT_BUTTON, self.onTestEndSound)

	def postInit(self):
		self.reportStartCheckBox.SetFocus()

	def onTestStartSound(self, evt):
		tones.beep(self.startHzEdit.GetValue(), self.startLengthEdit.GetValue())

	def onTestEndSound(self, evt):
		tones.beep(self.endHzEdit.GetValue(), self.endLengthEdit.GetValue())

	def onSave(self):
		config.conf["cursorLocator"]["reportStartOfLine"] = self.reportStartCheckBox.GetValue()
		config.conf["cursorLocator"]["reportLineLength"] = self.LengthEdit.GetValue()
		config.conf["cursorLocator"]["startLinePitch"] = self.startHzEdit.GetValue()
		config.conf["cursorLocator"]["startLineLength"] = self.startLengthEdit.GetValue()
		config.conf["cursorLocator"]["endLinePitch"] = self.endHzEdit.GetValue()
		config.conf["cursorLocator"]["endLineLength"] = self.endLengthEdit.GetValue()

class GlobalPlugin(globalPluginHandler.GlobalPlugin):

	scriptCategory = SCRCAT_SYSTEMCARET

	def __init__(self):
		super(globalPluginHandler.GlobalPlugin, self).__init__()
		NVDASettingsDialog.categoryClasses.append(AddonSettingsPanel)

	def terminate(self):
		NVDASettingsDialog.categoryClasses.remove(AddonSettingsPanel)

	def onSettings(self, evt):
		gui.mainFrame._popupSettingsDialog(NVDASettingsDialog, AddonSettingsPanel)

	@script(
		# Translators: Message presented in input mode.
		description=_("Shows the {} settings.").format(ADDON_SUMMARY),
		category=SCRCAT_CONFIG
	)
	def script_settings(self, gesture):
		wx.CallAfter(self.onSettings, None)

	def removeCarriageReturn(self, text):
		try:
			if ord(text[-1]) == 13 or ord(text[-1]) == 10:
				text = text[:-1]
			if ord(text[-1]) == 13 or ord(text[-1]) == 10:
				text = text[:-1]
		except IndexError:
			pass
		return text

	def event_typedCharacter(self, obj, nextHandler, ch):
		nextHandler()
		states = obj.states
		if not controlTypes.STATE_MULTILINE in states or controlTypes.STATE_READONLY in states:
			return
		if not ord(ch) >= 32:
			return
		reportStart = config.conf["cursorLocator"]["reportStartOfLine"]
		reportLineLength = config.conf["cursorLocator"]["reportLineLength"]
		startPitch = config.conf["cursorLocator"]["startLinePitch"]
		startLength = config.conf["cursorLocator"]["startLineLength"]
		endPitch = config.conf["cursorLocator"]["endLinePitch"]
		endLength = config.conf["cursorLocator"]["endLineLength"]
		if not reportStart and reportLineLength == 0:
			return
		try:
			info=obj.makeTextInfo(textInfos.POSITION_CARET)
			info.expand(textInfos.UNIT_LINE)
			text = self.removeCarriageReturn(info.text)
			if reportStart and len(text) == 1:
				tones.beep(startPitch, startLength)
			if reportLineLength > 0 and len(text) == reportLineLength:
				tones.beep(endPitch, endLength)
		except Exception as e:
			raise e

	@script(
		# Translators: Message presented in input help mode.
		description=_("Reports the length of the current line."),
		gesture="kb:NVDA+control+shift+l"
	)
	def script_reportLineLength(self, gesture):
		obj = api.getFocusObject()
		ti = obj.treeInterceptor
		if isinstance(ti, treeInterceptorHandler.DocumentTreeInterceptor) and not ti.passThrough:
			obj = ti
		try:
			info = obj.makeTextInfo(textInfos.POSITION_CARET)
			info.expand(textInfos.UNIT_LINE)
			lineLen = len(self.removeCarriageReturn(info.text))
			ui.message("Line length: %d" % lineLen)
		except Exception:
			pass
